### PR TITLE
fix(dnd): Add touch support to react-dnd

### LINKED ISSUE #3369: Mobile / touch support for fill in the gap with drag & drop
As reported by Quinn in Slack

### PR DESCRIPTION
**Pull request by CodingDive**
Fixes https://github.com/serlo/frontend/issues/3369, #3379

### CODE REVIEW

File: packages/editor/src/core/components/dnd-wrapper.tsx [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+import { ReactNode, useEffect, useState } from 'react'
 import { DndProvider } from 'react-dnd'
 import { HTML5Backend } from 'react-dnd-html5-backend'
+import { TouchBackend } from 'react-dnd-touch-backend'
 
 export const DndWrapper = ({ children }: { children: ReactNode }) => {
+  const [hasTouchScreen, setHasTouchScreen] = useState(false)
+
+  // ensure it runs on client (could the same be accomplished with 'use client'
+  // directive instead of using the effect?)
```
Comment (hejtful) [inline packages/editor/src/core/components/dnd-wrapper.tsx]: Could this help? https://nextjs.org/docs/app/api-reference/functions/userAgent

**Comment by Entkenntnis** [comment]
Opening https://frontend-git-fix-drag-drop-touch-support-serlo.vercel.app/mathe/298184/einhorn-der-mathematik-episode-1 on my android/chrome is not working yet

**Comment by hejtful** [comment]
I'll also work on fixing this now.

**Comment by hejtful** [comment]
Current version seems to work. Changes:
1. Lazy loaded the DndWrapper component without SSR, to avoid the hydration errors. Dragging on mobile still didn't work, so:
2. Refactored a bit to simplify logic and reduce room for errors. Not as neat as Mikey's code, but now it works.

**Comment by CodingDive** [comment]
https://frontend-chwwduwfk-serlo.vercel.app/mathe/298184/einhorn-der-mathematik-episode-1 works great now on Android and latest chrome. Only problem is that one doesn't see the items as it gets dragged. Only when it's dropped. Could someone check on an iPhone / Safari?

**Comment by hejtful** [comment]
Yeah, styling is not optimal, we could use [this package](https://louisbrunner.github.io/dnd-multi-backend/packages/react-dnd-preview/) to improve it. Edit: Just tried it, works only globally, so not really suitable for our use-case.

Sadly, I don't have an Apple mobile device to try it out.

**Comment by Entkenntnis** [comment]
DnD works for me now, too. The missing feedback (visible while dragging) is kind of a bummer, though 😢 

Another problem is that the touch backend has no mouse support. My device has touch enabled. The mouse is not working on my desktop.

**Comment by CodingDive** [comment]
> The mouse is not working on my desktop.

Yikes! Could you try again in a moment? I set `enableMouseEvents` to true; however there is a scary warning of the docs as I also wrote in the source code https://react-dnd.github.io/react-dnd/docs/backends/touch.

> NOTE: This is buggy due to the difference in touchstart/touchend event propagation compared to mousedown/mouseup/click.


If that is too buggy indeed, we could try https://github.com/LouisBrunner/dnd-multi-backend/tree/main as Vito suggested or another package to enable touch + mouse events.

**Comment by Entkenntnis** [comment]
@CodingDive This is how it looks like right now. It is still not working 😢 

firefox_k4vhhbaC5y

File: packages/editor/src/plugins/fill-in-the-blanks-exercise/renderer.tsx [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { DndWrapper } from '@editor/core/components/dnd-wrapper'
+import dynamic from 'next/dynamic'
```
Comment (elbotho) [inline packages/editor/src/plugins/fill-in-the-blanks-exercise/renderer.tsx]: using nextjs features in the editor breaks external integrations (e.g. edusharing)

**Review by elbotho — CHANGES_REQUESTED** [review]

**Comment by elbotho** [comment]
just noted: while dragging my cursor defaults to default cursor instead of dragging hand (mac, chrome and safari)

**Comment by CodingDive** [comment]
> while dragging my cursor defaults to default cursor instead of dragging hand (mac, chrome and safari)

Hmm some inconsistency from the OS. Works fine for me on Ubuntu with Firefox and Chrome. I added an extra line now, but if it still doesn't work, you could try and see where to add it.

File: packages/editor/src/plugins/fill-in-the-blanks-exercise/renderer.tsx (hunk 2 above)
Comment (hejtful) [reply packages/editor/src/plugins/fill-in-the-blanks-exercise/renderer.tsx]: Thanks, that was my bad...

**Comment by elbotho** [comment]
test results:
works on mac: chrome, safari, firefox and all also with emulated touch
also in android chrome
iOS: will test soon

**Review by Entkenntnis — APPROVED** [review]
Thank, the latest version is also working on my touch/mouse-device.

I think we can go again and merge this PR.

There are more usability considerations (e.g. see slack), but these can be handled in follow ups.

**Pull request merged**